### ISSUE TITLE
Parallel connection to multiple devices login failure 

### ISSUE THREAD

**Issue by renato-alves-barros** [CiscoTestAutomation/unicon.plugins#96]
Hello team,

I am using pyats/unicon to connect to a large testbed of devices in parallel using python threading library.

I have noticed that has I increase the parallelism of my connections, more and more devices fail to login.

The pattern is always the same:
Unicon matches the Escape character pattern `Escape character is '^]'.` and sends a couple of carriage returns. These carriage returns cause the Username to never be inserted on the router.
My understanding is that is should hold from sending those carriage returns because there is a Username pattern `Username:` which should match variable ESCAPE_CHAR_PROMPT_PATTERN:

[https://github.com/CiscoTestAutomation/unicon.plugins/blob/master/src/unicon/plugins/generic/settings.py](url)
```
        # Pattern to avoid sending 'enter' after Escape character pattern is seen
        self.ESCAPE_CHAR_PROMPT_PATTERN = r'.*(User Access Verification|sername:\s*$|assword:\s*$|login:\s*$|The highlighted entry will)'
```

However, maybe because of the parallelism, there is a context switch and the username pattern is seen later then it should and we end up sending the carriage returns, resulting in a failed authentication.

Is there any parameter that can help in this situation? Perhaps somehow delaying the carriage returns?

Logs are below:

```
Connected to 10.4.147.1.

Escape character is '^]'.

2024-03-04` 09:24:21,737: %UNICON-DEBUG: Statement matched: '[pattern='Escape character is .*\n', action=escape_char_callback, args=None, loop_continue=True, continue_timer=False, trim_buffer=True]'

2024-03-04 09:24:22,113: %UNICON-DEBUG: >>> Unicon Sending (target=PAN_ASR920_PEN_HUB):: '\r'

2024-03-04 09:24:22,666: %UNICON-DEBUG: >>> Unicon Sending (target=PAN_ASR920_PEN_HUB):: '\r'

Username:

2024-03-04 09:24:23,931: %UNICON-DEBUG: <<< Unicon Got (target=PAN_ASR920_PEN_HUB):: '\r\n\r\n\r\nUsername:'`
2024-03-04 09:24:23,982: %UNICON-DEBUG: *** Pattern Matched (target=PAN_ASR920_PEN_HUB):: '^.*([Uu]sername|[Ll]ogin): ?$'
2024-03-04 09:24:23,983: %UNICON-DEBUG: >>> Unicon Sending (target=PAN_ASR920_PEN_HUB):: 'Renato.Barros\r'
% Authentication failed
Username:
```

**Comment by SohanTirpude** [comment]
Hello @renato-alves-barros,

Kindly give me some time to analyze this issue.

Thank you.

**Comment by SohanTirpude** [comment]
Hello @renato-alves-barros,

It seems you raised this ticket [#226](https://github.com/CiscoTestAutomation/pyats/issues/226) in the past to connect to devices parallelly. 

So, rather than using the python's threading library, you can use the pyats in-built `tb.connect()` to connect all devices all available in the testbed through multiprocessing. 

Here is the example: https://github.com/CiscoTestAutomation/testing-bootcamp/blob/main/5-Scripts/nxos/8-parallel.py#L15, if you need.

Kindly try this way and let me know the status.

Thank you.

**Comment by SohanTirpude** [comment]
Hello @renato-alves-barros,

Is there any update on this?

Kindly let me know.

Thank you.

**Comment by renato-alves-barros** [comment]
But threading lib should work right?

Problem with testbed.connect() method is that if one device fails to connect, the full batch of devices fail to connect.
That's why I abandoned that solution.

Anyways, you can close the ticket, I played around with the following parameters and it improved:

ESCAPE_CHAR_CHATTY_TERM_WAIT=10,
ESCAPE_CHAR_PROMPT_WAIT=2,
ESCAPE_CHAR_PROMPT_WAIT_RETRIES=2

**Comment by SohanTirpude** [comment]
> Problem with testbed.connect() method is that if one device fails to connect, the full batch of devices fail to connect.
> That's why I abandoned that solution.
Are you saying that even after fix provide for https://github.com/CiscoTestAutomation/pyats/issues/226, you are still not able to connect to all devices?

**Comment by SohanTirpude** [comment]
Hello @renato-alves-barros,

If `testbed.connect()` is not working for you even with the latest pyats then kindly raise the ticket with proper details. Meanwhile, I am closing this issue as confirmed by you with few changes in configuration you were able to connect to devices. 
If you are facing any other issues, kindly open a new ticket.

Thank you.
